smart-engine/src/main/java/org/smartdata/server/engine/rule/SmallFilePlugin.java

```diff
@@ -0,0 +1,349 @@
+      this.containerFilePermission = containerFilePermission;
+      this.smartFiles = smartFiles;
+    }
+  }
+
+  /**
+   * Get valid compact action arguments.
+   */
+  private CompactActionArgs getCompactActionArgs(String firstFileDir,
+      SmartFilePermission firstFilePermission, List<String> smallFileList) {
+    if (!containerFileCache.isEmpty()) {
+      for (Map.Entry<String, ContainerFileInfo> entry
+          : containerFileCache.entrySet()) {
+        String containerFile = entry.getKey();
+        String containerFileDir = containerFile.substring(
+            0, containerFile.lastIndexOf("/") + 1);
+        SmartFilePermission containerFilePermission =
+            entry.getValue().smartFilePermission;
+        if (firstFileDir.equals(containerFileDir)
+            && firstFilePermission.equals(containerFilePermission)) {
+          List<String> validSmallFiles;
+          try {
+            validSmallFiles = getValidSmallFiles(containerFile, smallFileList);
+          } catch (MetaStoreException e) {
+            return genCompactActionArgs(firstFileDir,
+                firstFilePermission, smallFileList);
+          }
+          if (validSmallFiles != null) {
+            return new CompactActionArgs(containerFile, null, validSmallFiles);
+          }
+        }
+      }
+    }
+    return genCompactActionArgs(firstFileDir,
+        firstFilePermission, smallFileList);
+  }
+
+  /**
+   * Generate new compact action arguments based on first file info.
+   */
+  private CompactActionArgs genCompactActionArgs(String firstFileDir,
+      SmartFilePermission firstFilePermission, List<String> smallFileList) {
+    // Generate new container file
+    String containerFile = firstFileDir + CONTAINER_FILE_PREFIX
+        + UUID.randomUUID().toString().replace("-", "");
+    containerFileCache.put(containerFile,
+        new ContainerFileInfo(0, firstFilePermission));
+    return new CompactActionArgs(containerFile,
+        firstFilePermission, smallFileList);
+  }
+
+  /**
+   * Get valid small files according to container file.
+   */
+  private List<String> getValidSmallFiles(String containerFile,
+      List<String> smallFileList) throws MetaStoreException {
+    // Get container file len
+    FileInfo containerFileInfo;
+    long containerFileLen;
+    try {
+      containerFileInfo = metaStore.getFile(containerFile);
+    } catch (MetaStoreException e) {
+      LOG.error("Failed to get container file length: " + containerFile);
+      throw e;
+    }
+    if (containerFileInfo == null) {
+      containerFileCache.remove(containerFile);
+      return null;
+    } else {
+      containerFileLen = containerFileInfo.getLength();
+    }
+
+    // Get small files can be compacted to container file
+    List<String> ret = new ArrayList<>();
+    List<FileInfo> smallFileInfos = metaStore.getFilesByPaths(smallFileList);
+    for (FileInfo fileInfo : smallFileInfos) {
+      containerFileLen += fileInfo.getLength();
```

Comment:
Better to sort the files by length first, and add to `ret` from the shortest one.